crates/swc_ecma_transforms_base/src/dropped_comments_preserver.rs

```diff
@@ -0,0 +1,54 @@
+use swc_common::{comments::Comments, BytePos, Span};
+use swc_ecma_visit::{as_folder, noop_visit_mut_type, Fold, VisitMut, VisitMutWith};
+
+/// Preserves comments that would otherwise be dropped.
+///
+/// If during compilation an ast node associated with
+/// a comment is dropped, the comment will not appear in the final emitted
+/// output. This can create problems in the JavaScript ecosystem, particularly
+/// around instanbul coverage and other tooling that relies on comment
+/// directives.
+///
+/// This transformer shifts orphaned comments to the next closest ast-node.
+
+pub fn dropped_comments_preserver(comments: Option<&dyn Comments>) -> impl '_ + Fold + VisitMut {
+    as_folder(DroppedCommentsPreserver {
+        comments,
+        comment_position: 0,
+        is_first_span: true,
+    })
+}
+
+struct DroppedCommentsPreserver<'a> {
+    comments: Option<&'a dyn Comments>,
+    comment_position: u32,
+    is_first_span: bool,
+}
+
+impl VisitMut for DroppedCommentsPreserver<'_> {
+    noop_visit_mut_type!();
+
+    fn visit_mut_span(&mut self, span: &mut Span) {
+        if span.is_dummy() || self.is_first_span {
+            self.is_first_span = false;
+            return;
+        }
+
+        let mut leading_comments = Vec::new();
+        let mut trailing_comments = Vec::new();
+
+        for idx in (self.comment_position..=span.lo.0).map(BytePos) {
```

Comment:
Oh, this will be extremely slow.

Reply:
It removes whole advantage of storing comments in a shared hash map.

Reply:
@kdy1 Sorry, it was an extremely busy day for me. I'm just now sitting down and thinking about how to make this more this more performant. 

Reviewing the discord convo from last night, your suggestion to use `SingleThreadedComments` is unclear to me. I don't see a way to depart from the `SwcComments` data structure that is used top-level in `Compiler` without _replacing_ the compiler type.

This seems to be an appropriate way to solve this problem:

- Collect all candidate spans with a visitor
- iterate _exactly once_ over the `comments` entries, shifting orphaned comments to candidate spans

I understand the reasons that iterating over the concurrent hashmaps is undesirable, but reducing that iteration to one time outside of the visitor sounds like an acceptable concession. Do you agree?

If not, could you provide some guidance on how exactly you intend single threaded comments to be used.

Reply:
> because this needs to be implemented as a pass baked into existing compiler

No, it should not and it does not have to.
It will used inside `fn transform` and it will be dropped once `fn transform()` is finished.